sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/processor/CompactBlockProcessor.kt

```diff
@@ -670,6 +995,268 @@ class CompactBlockProcessor internal constructor(
+            limitTime: Long
+        ): Boolean {
+            return (currentTimeMillis - lastPreparationTime) >= limitTime
+        }
+
+        /**
+         * This operation fetches and returns the latest block height (chain tip)
+         *
+         * @return Latest block height wrapped in BlockHeight object, or null in case of failure
+         */
+        @VisibleForTesting
+        internal suspend fun fetchLatestBlockHeight(
+            downloader: CompactBlockDownloader,
+            network: ZcashNetwork
+        ): BlockHeight? {
+            Twig.debug { "Fetching latest block height..." }
+
+            var latestBlockHeight: BlockHeight? = null
+
+            retryUpToAndContinue(FETCH_LATEST_BLOCK_HEIGHT_RETRIES) {
+                when (val response = downloader.getLatestBlockHeight()) {
+                    is Response.Success -> {
+                        Twig.debug { "Latest block height fetched successfully with value: ${response.result.value}" }
+                        latestBlockHeight = runCatching {
+                            response.result.toBlockHeight(network)
+                        }.getOrNull()
+                    }
+                    is Response.Failure -> {
+                        Twig.error { "Fetching latest block height failed with: ${response.toThrowable()}" }
+                        throw LightWalletException.GetLatestBlockHeightException(
+                            response.code,
+                            response.description,
+                            response.toThrowable()
+                        )
+                    }
+                }
+            }
+
+            return latestBlockHeight
+        }
+
+        /**
+         * This operation downloads note commitment tree data from the lightwalletd server to decide if we communicate
+         * with linear or spend-before-sync node
+         *
+         * @return GetSubtreeRootsResult as a wrapper for the lightwalletd response result
+         */
+        @VisibleForTesting
+        internal suspend fun getSubtreeRoots(
+            downloader: CompactBlockDownloader,
+            network: ZcashNetwork
+        ): GetSubtreeRootsResult {
+            Twig.debug { "Fetching SubtreeRoots..." }
+
+            var result: GetSubtreeRootsResult = GetSubtreeRootsResult.Linear
+
+            retryUpToAndContinue(GET_SUBTREE_ROOTS_RETRIES) {
+                downloader.getSubtreeRoots(
+                    startIndex = 0,
+                    maxEntries = if (network.isTestnet()) {
+                        65536
+                    } else {
+                        0
+                    },
+                    shieldedProtocol = ShieldedProtocolEnum.SAPLING
+                ).onEach { response ->
+                    when (response) {
+                        is Response.Success -> {
+                            Twig.verbose {
+                                "SubtreeRoot got successfully: it's completingHeight: ${response.result
```

Comment:
```suggestion
                                "SubtreeRoot fetched successfully: its completingHeight is ${response.result
```
(possessive "its"). Non-blocking.